src/env/__init_tls.c

```diff
@@ -7,111 +7,155 @@
-	return 0;
-}
-
-int __init_utp(void *p, int set_tp)
-{
-	struct lthread_tcb_base *tcb = (struct lthread_tcb_base *)p;
-	tcb->self = p;
-	tcb->schedctx = __scheduler_self();
-	if (libc.user_tls_enabled && set_tp)
+	if (sgxlkl_enclave->mode == SW_DEBUG_MODE)
 	{
-		if (sgxlkl_enclave->mode == SW_DEBUG_MODE)
-		{
-			int r = __set_thread_area(TP_ADJ(p));
-			if (r < 0)
-			{
-				sgxlkl_fail("Could not set thread area %p: %s\n", p, strerror(errno));
-			}
-		}
-		else
+		int r = __set_thread_area(TP_ADJ(p));
+		if (r < 0)
 		{
-			__asm__ volatile("wrfsbase %0" ::"r"(p));
+			sgxlkl_fail("Could not set thread area %p: %s\n", p, strerror(errno));
 		}
 	}
+	else
+	{
+		__asm__ volatile("wrfsbase %0" ::"r"(p));
+	}
+	//int r = __set_thread_area(TP_ADJ(p));
+	//if (r < 0) return -1;
+	// if (!r) 
+	libc.can_do_threads = 1;
+	td->detach_state = DT_JOINABLE;
+	td->tid = __syscall(SYS_set_tid_address, &td->detach_state);
+	td->locale = &libc.global_locale;
+	td->robust_list.head = &td->robust_list.head;
 	return 0;
 }
 
-static struct builtin_tls
-{
+static struct builtin_tls {
 	char c;
-	struct schedctx *pt;
+	struct pthread pt;
 	void *space[16];
 } builtin_tls[1];
 #define MIN_TLS_ALIGN offsetof(struct builtin_tls, pt)
 
 static struct tls_module main_tls;
 
-/* Set up user-level thread TLS */
-void *__copy_utls(struct lthread *lt, unsigned char *mem, size_t sz)
+void *__copy_tls(unsigned char *mem)
 {
+	pthread_t td;
 	struct tls_module *p;
 	size_t i;
 	uintptr_t *dtv;
 
+#ifdef TLS_ABOVE_TP
+	dtv = (uintptr_t*)(mem + libc.tls_size) - (libc.tls_cnt + 1);
+
+	mem += -((uintptr_t)mem + sizeof(struct pthread)) & (libc.tls_align-1);
+	td = (pthread_t)mem;
+	mem += sizeof(struct pthread);
+
+	for (i=1, p=libc.tls_head; p; i++, p=p->next) {
+		dtv[i] = (uintptr_t)(mem + p->offset) + DTP_OFFSET;
+		memcpy(mem + p->offset, p->image, p->len);
+	}
+#else
 	dtv = (uintptr_t *)mem;
 
-	mem += sz - sizeof(struct lthread_tcb_base);
-	mem -= (uintptr_t)mem & (libc.tls_align - 1);
+	mem += libc.tls_size - sizeof(struct pthread);
+	mem -= (uintptr_t)mem & (libc.tls_align-1);
+	td = (pthread_t)mem;
 
-	for (i = 1, p = libc.tls_head; p; i++, p = p->next)
-	{
+	for (i=1, p=libc.tls_head; p; i++, p=p->next) {
 		dtv[i] = (uintptr_t)(mem - p->offset) + DTP_OFFSET;
 		memcpy(mem - p->offset, p->image, p->len);
 	}
+#endif
 	dtv[0] = libc.tls_cnt;
-	lt->dtv = lt->dtv_copy = dtv;
-	return (void *)mem;
+	td->dtv = td->dtv_copy = dtv;
+	return td;
 }
 
-/* Initialisation of user-level thread TLS image */
-void __init_utls(struct tls_module *apptls)
+#if ULONG_MAX == 0xffffffff
+typedef Elf32_Phdr Phdr;
+#else
+typedef Elf64_Phdr Phdr;
+#endif
+
+extern weak hidden const size_t _DYNAMIC[];
+
+static void static_init_tls(size_t *aux)
 {
-	if (apptls && apptls->image)
-	{
-		main_tls = *apptls;
+	unsigned char *p;
+	size_t n;
+	Phdr *phdr, *tls_phdr=0;
+	size_t base = 0;
+	void *mem;
+
+	for (p=(void *)aux[AT_PHDR],n=aux[AT_PHNUM]; n; n--,p+=aux[AT_PHENT]) {
```

Comment:
Does this work? I’m not sure we provide a sufficiently compatible initial stack layout for this.  We should be putting a few things in the elf aux args area including the initial thread size, the pointer to the syscall function, and the attestation quote.

Reply:
This is the weak version of `__init_tls` used in static linking. For dynamically loaded apps it should be superseded by the empty definition in `ldso/dynlink.c`.

Reply:
Thanks.  We will want to support static loading soon so that we can test the binfmt code, but it's not urgent yet.